### ISSUE TITLE
Expressions

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#9803]
**tl:dr** ticket discussing public API design for expressions [#9439](https://github.com/mapbox/mapbox-gl-native/pull/9439) on Android. 

## Background

With Android SDK 5.0.0, we introduced [Functions](https://www.mapbox.com/mapbox-gl-js/style-spec/#types-function) to support data driven styling:

>The value for any layout or paint property may be specified as a function. Functions allow you to make the appearance of a map feature change with the current zoom level and/or the feature's properties.

For example, the Android binding exposes an interval zoom function as

```java
    layer.setProperties(
      fillColor(
        zoom(
          interval(
            stop(1, fillColor(Color.RED)),
            stop(5, fillColor(Color.BLUE)),
            stop(10, fillColor(Color.GREEN))
          )
        )
      )
    );
```

This will change the `fillColor` of `FillLayer` depending on the zoom level of a map.

## Way forward

[Project Arbex](https://github.com/mapbox/mapbox-gl-js/pull/4777) in gl-js introduces the concept of using expressions as part of Functions. 

> Required expression value An expression defines how one or more feature property values and/or the current zoom level are **combined using logical, mathematical, string, or color operations to produce the appropriate style value**.

See the [style spec draft](https://github.com/mapbox/mapbox-gl-js/blob/9ec03f04d9299cc09cce55ceb9957c67bd5f104b/docs/style-spec/expressions.md#Expressions)  of functions for more details.

## API proposal

On the Android binding we would like to continue the DSL approach taken with the initial Function implementation. We still need to identify what it will take to bring the expression concept to the Android binding. TBD.
